File: backend/modules/discovery/cidr_range.ts
/**
 * Expansão de faixas CIDR IPv4 para a varredura de descoberta.
 *
 * Vive fora dos scanners porque três lugares precisam da mesma resposta: o
 * scanner ICMP (quais IPs pingar), o endpoint que dispara a varredura de uma
 * rede (o CIDR cadastrado é utilizável?) e a UI (quantos hosts serão varridos).
 */

/** Teto de endereços varridos por execução — /22 completo. */
export const MAX_SCAN_HOSTS = 1024

export interface CidrRange {
  /** Endereço de rede normalizado, ex.: `192.168.1.0` */
  networkAddress: string
  prefix: number
  /** Total de endereços utilizáveis na faixa, antes de qualquer truncamento */
  usableHosts: number
  /** `true` quando `usableHosts` excede `MAX_SCAN_HOSTS` */
  truncated: boolean
}

export class InvalidCidrError extends Error {
  constructor(cidr: string, reason: string) {
    super(`Faixa CIDR inválida "${cidr}": ${reason}`)
    this.name = 'InvalidCidrError'
  }
}

function toNumber(ip: string): number | null {
  const parts = ip.split('.')
  if (parts.length !== 4) return null

  let value = 0
  for (const part of parts) {
    if (!/^\d{1,3}$/.test(part)) return null
    const octet = Number(part)
    if (octet > 255) return null
    // `<<` opera em 32 bits com sinal; a multiplicação evita o valor negativo
    // que apareceria no primeiro octeto acima de 127.
    value = value * 256 + octet
  }

  return value
}

function toAddress(value: number): string {
  return [
    Math.floor(value / 16_777_216) % 256,
    Math.floor(value / 65_536) % 256,
    Math.floor(value / 256) % 256,
    value % 256,
  ].join('.')
}

/**
 * Interpreta e valida uma faixa. Aceita host único (sem `/`) e prefixos de
 * /8 a /32 — abaixo de /8 a varredura deixa de fazer sentido para o alvo do
 * produto (redes residenciais e de pequenas empresas).
 */
export function parseCidrRange(cidr: string): CidrRange {
  const value = (cidr ?? '').trim()
  if (!value) throw new InvalidCidrError(cidr, 'valor vazio')

  const [address, prefixPart] = value.split('/')
  const base = toNumber(address.trim())
  if (base === null) throw new InvalidCidrError(cidr, 'endereço IPv4 malformado')

  // Host avulso: uma varredura de um endereço só é legítima (testar um alvo).
  const prefix = prefixPart === undefined ? 32 : Number(prefixPart)
  if (!Number.isInteger(prefix) || prefix < 8 || prefix > 32) {
    throw new InvalidCidrError(cidr, 'prefixo deve estar entre /8 e /32')
  }

  const size = 2 ** (32 - prefix)
  const networkNumber = Math.floor(base / size) * size

  // /31 e /32 não têm endereço de rede nem de broadcast reservados (RFC 3021)
  const usableHosts = prefix >= 31 ? size : size - 2

  return {
    networkAddress: toAddress(networkNumber),
    prefix,
    usableHosts,
    truncated: usableHosts > MAX_SCAN_HOSTS,
  }
}

/**
 * Endereços a varrer na faixa, já sem rede e broadcast.
 *
 * Faixas maiores que `MAX_SCAN_HOSTS` são truncadas no início do bloco — varrer
 * um /16 inteiro (65 mil pings) travaria o probe por horas. O truncamento é
 * visível em `parseCidrRange().truncated` para a UI poder avisar em vez de
 * silenciosamente varrer só um pedaço.
 */
export function expandCidr(cidr: string, maxHosts = MAX_SCAN_HOSTS): string[] {
  const range = parseCidrRange(cidr)
  const size = 2 ** (32 - range.prefix)
  const networkNumber = toNumber(range.networkAddress)!

  const first = range.prefix >= 31 ? networkNumber : networkNumber + 1
  const last = range.prefix >= 31 ? networkNumber + size - 1 : networkNumber + size - 2
  const limit = Math.min(last, first + maxHosts - 1)

  const addresses: string[] = []
  for (let current = first; current <= limit; current++) {
    addresses.push(toAddress(current))
  }

  return addresses
}

/** `true` se o CIDR é utilizável numa varredura. */
export function isScannableCidr(cidr: string): boolean {
  try {
    parseCidrRange(cidr)
    return true
  } catch {
    return false
  }
}

File: backend/src/services/discovery/cidr_range.rs
//! Expansão de faixas CIDR para a varredura de descoberta.
//!
//! Reexporta e delega para o módulo unificado [`crate::services::shared::cidr`].

use std::net::IpAddr;

use crate::services::shared::{
    cidr::{self, DiscoveryCidrRange},
    errors::AppError,
};

/// Tamanho de cada lote. A execução percorre todos os lotes do CIDR.
pub const MAX_SCAN_HOSTS: u32 = cidr::MAX_SCAN_HOSTS;

pub type CidrRange = DiscoveryCidrRange;

/// Interpreta e valida uma faixa. Aceita host único (sem `/`) e prefixos de
/// /8 a /32 ou IPv6 de /112 a /128.
pub fn parse_cidr_range(cidr: &str) -> Result<CidrRange, AppError> {
    cidr::parse_discovery_cidr(cidr)
}

/// `true` se o CIDR é utilizável numa varredura.
#[must_use]
pub fn is_scannable_cidr(cidr: &str) -> bool {
    cidr::is_scannable_cidr(cidr)
}

/// Expande somente endereços utilizáveis e limita a memória/tempo da operação.
pub fn expand_cidr(cidr: &str, max_hosts: usize) -> Result<Vec<IpAddr>, AppError> {
    cidr::expand_cidr(cidr, max_hosts)
}

/// Expande um lote do CIDR sem materializar a faixa inteira em memória.
pub fn expand_cidr_batch(
    cidr: &str,
    offset: u32,
    max_hosts: usize,
) -> Result<Vec<IpAddr>, AppError> {
    cidr::expand_cidr_batch(cidr, offset, max_hosts)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normaliza_o_endereco_de_rede() {
        let range = parse_cidr_range("192.168.1.77/24").unwrap();
        assert_eq!(range.network_address, "192.168.1.0");
        assert_eq!(range.prefix, 24);
        assert_eq!(range.usable_hosts, 254);
        assert!(!range.truncated);
    }

    #[test]
    fn host_avulso_vira_prefixo_32() {
        let range = parse_cidr_range("10.0.0.9").unwrap();
        assert_eq!(range.prefix, 32);
        assert_eq!(range.usable_hosts, 1);
        assert_eq!(range.network_address, "10.0.0.9");
    }

    #[test]
    fn rfc_3021_nao_reserva_rede_e_broadcast() {
        assert_eq!(parse_cidr_range("10.0.0.0/31").unwrap().usable_hosts, 2);
        assert_eq!(parse_cidr_range("10.0.0.0/32").unwrap().usable_hosts, 1);
        assert_eq!(parse_cidr_range("10.0.0.0/30").unwrap().usable_hosts, 2);
    }

    #[test]
    fn faixas_grandes_nao_sao_mais_truncadas() {
        assert!(!parse_cidr_range("10.0.0.0/22").unwrap().truncated);
        assert!(!parse_cidr_range("10.0.0.0/21").unwrap().truncated);
    }

    #[test]
    fn primeiro_octeto_acima_de_127_nao_vira_negativo() {
        let range = parse_cidr_range("200.160.0.0/24").unwrap();
        assert_eq!(range.network_address, "200.160.0.0");
    }

    #[test]
    fn recusa_entrada_malformada_com_a_mensagem_do_backend_atual() {
        for (cidr, trecho) in [
            ("", "valor vazio"),
            ("   ", "valor vazio"),
            ("192.168.1", "endereço IP malformado"),
            ("192.168.1.256/24", "endereço IP malformado"),
            ("192.168.1.1.1/24", "endereço IP malformado"),
            ("abc/24", "endereço IP malformado"),
            ("10.0.0.0/7", "prefixo deve estar entre /8 e /32"),
            ("10.0.0.0/33", "prefixo deve estar entre /8 e /32"),
            ("10.0.0.0/abc", "prefixo deve estar entre /8 e /32"),
        ] {
            let err = parse_cidr_range(cidr).unwrap_err();
            let msg = err.to_string();
            assert!(
                msg.contains(trecho),
                "`{cidr}` devia falhar com `{trecho}`, veio `{msg}`"
            );
            assert!(!is_scannable_cidr(cidr));
        }
    }

    #[test]
    fn faixas_validas_sao_varreduraveis() {
        for cidr in ["192.168.0.0/24", "10.0.0.0/8", "172.16.5.9", "10.0.0.0/32"] {
            assert!(is_scannable_cidr(cidr), "`{cidr}` devia ser varredurável");
        }
    }

    #[test]
    fn expansao_respeita_rfc_3021_e_limite() {
        assert_eq!(
            expand_cidr("10.0.0.0/31", 1024).unwrap(),
            vec![
                "10.0.0.0".parse::<IpAddr>().unwrap(),
                "10.0.0.1".parse::<IpAddr>().unwrap()
            ]
        );
        assert_eq!(expand_cidr("192.168.1.0/24", 2).unwrap().len(), 2);
    }

    #[test]
    fn ipv6_e_normalizado_e_expandido_em_lotes() {
        let range = parse_cidr_range("fd00::7/126").unwrap();
        assert_eq!(range.network_address, "fd00::4");
        assert_eq!(range.usable_hosts, 4);
        assert_eq!(
            expand_cidr_batch("fd00::7/126", 2, 2).unwrap(),
            vec![
                "fd00::6".parse::<IpAddr>().unwrap(),
                "fd00::7".parse().unwrap()
            ]
        );
        assert!(parse_cidr_range("fd00::/64").is_err());
    }
}
